---
layout: page
title: About
nav_order: 1
---


<h1 dir="auto" style="border-bottom: 1px solid grey; padding-bottom: 10px; box-sizing: border-box;">
    <img src="assets/images/logo.png" style="margin-bottom: 0px; align: right; height: 3rem; max-width: 100%; float: left; padding: 0px 15px 3px 15px">
    tidyproteomics-interactive
</h1>

This is the documentation page for [**tidyproteomics-interactive**](https://github.com/ejmackrell/tidyproteomics-interactive), a Shiny application for interactively working with the R package [**tidyproteomics**](https://github.com/jeffsocal/tidyproteomics). Users may upload their protein- or peptide-level data for abundance subsetting, contaminant removal, abundance normalization, differential expression analysis, and ontology enrichment.
<br><br>
<img class="preview" src="assets/images/preview-bg.png">
<br><br>
If you use this application for your research, please cite [our article](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-023-05360-7).

>Jones, J., MacKrell, E.J., Wang, TY. *et al*. Tidyproteomics: an open-source R package and data object for quantitative proteomics post analysis and visualization. *BMC Bioinformatics* **24**, 239 (2023).